File: agente/clients/transcrever.py
"""Cliente para o servidor proprio de transcricao (faster-whisper no VPS Minerador SaaS)."""
import logging
import uuid

import httpx

from config import settings

logger = logging.getLogger(__name__)


_MIME_EXTENSIONS = {
    "audio/ogg": "ogg",
    "audio/ogg; codecs=opus": "ogg",
    "audio/mpeg": "mp3",
    "audio/mp4": "m4a",
    "audio/aac": "aac",
    "audio/wav": "wav",
    "audio/x-wav": "wav",
    "audio/webm": "webm",
}


def _extension_for(mime_type: str) -> str:
    if not mime_type:
        return "ogg"
    base = mime_type.split(";", 1)[0].strip().lower()
    return _MIME_EXTENSIONS.get(base) or _MIME_EXTENSIONS.get(mime_type.lower(), "ogg")


async def transcribe(audio_bytes: bytes, mime_type: str = "audio/ogg") -> str:
    """Envia audio para o servidor proprio e retorna o texto transcrito.

    Pipeline:
    1. POST /videos/upload com multipart/form-data, retorna {video_path}.
    2. POST /transcribe com {video_path, language: 'pt'}, retorna o texto.
    """
    base_url = settings.TRANSCREVER_API_URL.rstrip("/")
    if not base_url:
        logger.error("TRANSCREVER_API_URL nao configurado")
        return ""

    filename = f"wa_audio_{uuid.uuid4().hex}.{_extension_for(mime_type)}"

    try:
        async with httpx.AsyncClient(timeout=120.0) as client:
            upload_resp = await client.post(
                f"{base_url}/videos/upload",
                files={"file": (filename, audio_bytes, mime_type or "application/octet-stream")},
            )
            if upload_resp.status_code != 200:
                logger.error(
                    "transcrever upload falhou status=%s body=%s",
                    upload_resp.status_code,
                    upload_resp.text[:300],
                )
                return ""
            upload_data = upload_resp.json()
            video_path = (
                upload_data.get("video_path")
                or upload_data.get("path")
                or upload_data.get("file_path")
                or upload_data.get("filename")
            )
            if not video_path:
                logger.error("upload sem caminho retornado: %s", upload_data)
                return ""

            tx_resp = await client.post(
                f"{base_url}/transcribe",
                json={"video_path": video_path, "language": "pt"},
            )
            if tx_resp.status_code != 200:
                logger.error(
                    "transcrever /transcribe falhou status=%s body=%s",
                    tx_resp.status_code,
                    tx_resp.text[:300],
                )
                return ""
            data = tx_resp.json()
            text = (
                data.get("text")
                or data.get("transcription")
                or data.get("transcricao")
                or data.get("result")
                or ""
            )
            if not text and isinstance(data, dict):
                segments = data.get("segments") or []
                if isinstance(segments, list):
                    text = " ".join(seg.get("text", "") for seg in segments if isinstance(seg, dict)).strip()
            return text.strip() if isinstance(text, str) else ""
    except Exception as exc:
        logger.error("transcrever erro: %s", exc, exc_info=True)
        return ""

File: src/lib/clients/transcrever.ts
/**
 * Cliente do servidor proprio de transcricao (Whisper local no VPS).
 *
 * Contrato confirmado pelo servidor real (api.transcrever, "Video Transcription
 * API" na porta 8765 atras de nginx na porta 80) e pelo pipeline de referencia
 * em mvp_agent_vibe/audio/pipeline.py:
 *   1. POST {BASE}/videos/upload (multipart) -> { filename }
 *   2. POST {BASE}/transcribe { video_path: "/opt/transcrever/videos/<filename>",
 *      language: "pt" } -> { status: "success", text }
 *
 * Observacao: o servidor responde por HTTP (porta 80), nao HTTPS (sem SSL na 443).
 * Por isso o default e http://. Configuravel por TRANSCREVER_API_URL. Zero token de LLM.
 */
const BASE_URL = (
  process.env.TRANSCREVER_API_URL ?? "http://api.transcrever.seu-dominio.com"
).replace(/\/$/, "");

// Path interno do servidor onde o upload salva os arquivos.
const VIDEOS_PATH = (
  process.env.TRANSCREVER_VIDEOS_PATH ?? "/opt/transcrever/videos"
).replace(/\/$/, "");

const TIMEOUT_MS = 180_000;

export async function transcribeAudio(
  audio: Uint8Array,
  mimeType = "audio/ogg",
): Promise<string> {
  if (!BASE_URL) return "";

  const suffix = mimeType.includes("ogg") ? "ogg" : "mp4";
  const filename = `wa_${crypto.randomUUID().replace(/-/g, "").slice(0, 30)}.${suffix}`;

  try {
    // 1. Upload
    const form = new FormData();
    form.append(
      "file",
      new Blob([audio as BlobPart], { type: mimeType || "application/octet-stream" }),
      filename,
    );
    const upload = await fetch(`${BASE_URL}/videos/upload`, {
      method: "POST",
      body: form,
      signal: AbortSignal.timeout(TIMEOUT_MS),
    });
    if (!upload.ok) {
      console.error(`[transcrever] upload falhou: ${upload.status}`);
      return "";
    }
    const upData = (await upload.json()) as { filename?: string };
    const uploadedFilename = upData.filename ?? filename;

    // 2. Transcrever com o caminho absoluto do servidor
    const videoPath = `${VIDEOS_PATH}/${uploadedFilename}`;
    const tx = await fetch(`${BASE_URL}/transcribe`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ video_path: videoPath, language: "pt" }),
      signal: AbortSignal.timeout(TIMEOUT_MS),
    });
    if (!tx.ok) {
      console.error(`[transcrever] transcribe falhou: ${tx.status}`);
      return "";
    }
    const data = (await tx.json()) as { status?: string; text?: string };
    if (data.status !== "success") {
      console.error("[transcrever] status nao-success:", data.status);
      return "";
    }
    return (data.text ?? "").trim();
  } catch (err) {
    console.error("[transcrever] erro:", err);
    return "";
  }
}
